### ISSUE TITLE
Add documentation for .buckjavaargs

### ISSUE THREAD

**Issue by joshzana** [facebook/buck#300]
`.buckjavaargs` is completely undocumented, which makes it hard to discover how to change things like maximum memory.

For instance, Gradle supports this::

    dexOptions {
        javaMaxHeapSize "2g"
    }

This option is documented here: http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Advanced-Build-Customization

This is especially problematic for Buck since we run dx in process for faster builds.

**Comment by sdwilsh** [comment]
You can create a `.buckjavaargs` file next to your `.buckconfig` that contains arguments to Java.  In your case, you'd want it to contain `-Xmx2g`.

I'm going to morph this issue into documenting that.

**Comment by joshzana** [comment]
Thanks!  Does .buckjavaargs get used for program invocations?  It looks like it does not:

write_proguard_command_line_parameters
java -Xmx1024M -jar /Users/joshzana/Library/Android/sdk/tools/proguard/lib/proguard.jar ...

With the .buckjavaargs I get past dx, but then crash in proguard.

**Comment by sdwilsh** [comment]
Interesting.  Want to make a pull request to add an option for controlling the memory of invoking proguard too?  If not, please file a new issue for that.

**Comment by k21** [comment]
You can specify max heap size for proguard by putting this into your .buckconfig:
```
[tools]
proguard-max-heap-size = 2G
```

I am going to update the issue to add this to the documentation as well.

**Comment by joshzana** [comment]
Nice!  That got me past proguard, but not dx: 

UNEXPECTED TOP-LEVEL ERROR:
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at com.android.dx.dex.code.RopTranslator$TranslationVisitor.visitPlainInsn(RopTranslator.java:551)
	at com.android.dx.dex.code.RopTranslator$LocalVariableAwareTranslationVisitor.visitPlainInsn(RopTranslator.java:835)
	at com.android.dx.rop.code.PlainInsn.accept(PlainInsn.java:80)
	at com.android.dx.rop.code.InsnList.forEach(InsnList.java:76)
	at com.android.dx.dex.code.RopTranslator.outputBlock(RopTranslator.java:262)
	at com.android.dx.dex.code.RopTranslator.outputInstructions(RopTranslator.java:233)
	at com.android.dx.dex.code.RopTranslator.translateAndGetResult(RopTranslator.java:212)
	at com.android.dx.dex.code.RopTranslator.translate(RopTranslator.java:105)
	at com.android.dx.dex.cf.CfTranslator.processMethods(CfTranslator.java:317)
	at com.android.dx.dex.cf.CfTranslator.translate0(CfTranslator.java:137)
	at com.android.dx.dex.cf.CfTranslator.translate(CfTranslator.java:93)
	at com.android.dx.command.dexer.Main.processClass(Main.java:729)
	at com.android.dx.command.dexer.Main.processFileBytes(Main.java:673)
	at com.android.dx.command.dexer.Main.access$300(Main.java:83)
	at com.android.dx.command.dexer.Main$1.processFileBytes(Main.java:602)
	at com.android.dx.cf.direct.ClassPathOpener.processArchive(ClassPathOpener.java:284)
	at com.android.dx.cf.direct.ClassPathOpener.processOne(ClassPathOpener.java:166)
	at com.android.dx.cf.direct.ClassPathOpener.process(ClassPathOpener.java:144)
	at com.android.dx.command.dexer.Main.processOne(Main.java:632)
	at com.android.dx.command.dexer.Main.processAllFiles(Main.java:510)
	at com.android.dx.command.dexer.Main.runMonoDex(Main.java:280)
	at com.android.dx.command.dexer.Main.run(Main.java:246)
	at com.android.dx.command.dexer.Main.main(Main.java:215)
	at com.android.dx.command.Main.main(Main.java:106)


Seems to me there should be one config on .buckconfig that controls all of these, instead of a setting per tool.

**Comment by k21** [comment]
Hm, the heap size for the external dx is currently not customizable. To unblock yourself, you can override it in [DxStep](https://github.com/facebook/buck/blob/4be32ee3f68b5348ac44e39c042a835b4f5575cb/src/com/facebook/buck/android/DxStep.java#L43) by putting e.g. "-JXmx1g" there. We might make it configurable or at least increase the default in the future, but sending a pull request would be the fastest way to get the change in.

**Comment by joshzana** [comment]
I'm happy to fix, but it might take me a couple days to get to it.

**Comment by sdwilsh** [comment]
@joshzana - if you can't get to it, could you file a new issue about not being able to control that?

**Comment by joshzana** [comment]
@sdwilsh  thanks for doing that for me!
